Index: nginx-1.22.0/src/event/ngx_event_openssl.c
===================================================================
--- nginx-1.22.0.orig/src/event/ngx_event_openssl.c
+++ nginx-1.22.0/src/event/ngx_event_openssl.c
@@ -4690,6 +4690,26 @@ ngx_ssl_get_protocol(ngx_connection_t *c
     return NGX_OK;
 }
 
+ngx_int_t
+ngx_ssl_get_rtt(ngx_connection_t *c, ngx_pool_t *pool, ngx_str_t *s)
+{
+    long rtt = SSL_get_rtt(c->ssl->connection);
+    if (rtt < 1)
+        return NGX_ERROR;
+
+    u_char buf[100];
+    sprintf((char *)buf, "%ld", rtt);
+    size_t len = ngx_strlen(buf);
+
+    s->len = len;
+    s->data = ngx_pnalloc(pool, len);
+
+    if (s->data == NULL)
+        return NGX_ERROR;
+    ngx_memcpy(s->data, buf, len);
+
+    return NGX_OK;
+}
 
 ngx_int_t
 ngx_ssl_get_cipher_name(ngx_connection_t *c, ngx_pool_t *pool, ngx_str_t *s)
Index: nginx-1.22.0/src/event/ngx_event_openssl.h
===================================================================
--- nginx-1.22.0.orig/src/event/ngx_event_openssl.h
+++ nginx-1.22.0/src/event/ngx_event_openssl.h
@@ -252,6 +252,8 @@ ngx_int_t ngx_ssl_check_host(ngx_connect
 
 ngx_int_t ngx_ssl_get_protocol(ngx_connection_t *c, ngx_pool_t *pool,
     ngx_str_t *s);
+ngx_int_t ngx_ssl_get_rtt(ngx_connection_t *c, ngx_pool_t *pool,
+    ngx_str_t *s);
 ngx_int_t ngx_ssl_get_cipher_name(ngx_connection_t *c, ngx_pool_t *pool,
     ngx_str_t *s);
 ngx_int_t ngx_ssl_get_ciphers(ngx_connection_t *c, ngx_pool_t *pool,
Index: nginx-1.22.0/src/http/modules/ngx_http_ssl_module.c
===================================================================
--- nginx-1.22.0.orig/src/http/modules/ngx_http_ssl_module.c
+++ nginx-1.22.0/src/http/modules/ngx_http_ssl_module.c
@@ -361,6 +361,9 @@ static ngx_http_variable_t  ngx_http_ssl
     { ngx_string("ssl_server_name"), NULL, ngx_http_ssl_variable,
       (uintptr_t) ngx_ssl_get_server_name, NGX_HTTP_VAR_CHANGEABLE, 0 },
 
+    { ngx_string("ssl_rtt"), NULL, ngx_http_ssl_variable,
+      (uintptr_t) ngx_ssl_get_rtt, NGX_HTTP_VAR_CHANGEABLE, 0 },
+
     { ngx_string("ssl_alpn_protocol"), NULL, ngx_http_ssl_variable,
       (uintptr_t) ngx_ssl_get_alpn_protocol, NGX_HTTP_VAR_CHANGEABLE, 0 },
 
